### PR TITLE
Teacher dashboard experiment

### PR DESCRIPTION
**Pull request by maddiedierker**
### What it does
Below describes how the split test for determining which teachers see the old/new teacher dashboard will work based on where the teacher is coming from:

#### Dashboard (New) Teacher Dashboard
Every time a teacher visits the new teacher dashboard, we will check the value of our split test percentage (0-100). If the value is 0, and therefore the experiment is disabled, the teacher is redirected to the corresponding route in the pegasus/old teacher dashboard. Otherwise, the teacher is allowed to proceed and the `teacher-dashboard-react` experiment is officially enabled.

#### Pegasus (Old) Teacher Dashboard
Every time a teacher visits the old teacher dashboard, we will check the value of our split test percentage (0-100). If the value is 0, and therefore the experiment is disabled, we do nothing and allow the teacher to stay on the old teacher dashboard.

Next, we check the state of `teacher-dashboard-experiment` in local storage. It's value will be one of: "on", "off", or undefined. Based on that value, we do the following:
- **on:** Redirect the user to the corresponding route in the dashboard/new teacher dashboard, which enables the `teacher-dashboard-react` experiment
- **off:** Allow the user to continue using the old teacher dashboard
- **undefined:** Randomly set the value of `teacher-dashboard-experiment` in local storage using our split test percentage (more specifically: `turnOnExperiment = Math.floor(Math.random() * 100) <= splitTestPercentage`). Then, `teacher-dashboard-experiment` is set to either "on" or "off" and we proceed as described above.

### Diagram
View the original drawing [here](https://docs.google.com/drawings/d/1PayLR_SjXlvf7mig7GSwhcEkoi6C_xRtjJRoMJ9nTQg/edit?usp=sharing).
untitled drawing

### CODE REVIEW

**Comment by maddiedierker** [comment]
re: unit tests for this script -- i couldn't find where we have any javascript unit tests in pegasus, so i'm not sure the best way to set them up. i'd like to add a few unit tests (mostly to test my regex skills...), but this i'm also okay with shipping this since it can very easily be turned off 🤷‍♀️

**Comment by davidsbailey** [comment]
Hi Maddie, nice work, this looks like a nice and creative solution to the problem! I think the part I'm missing here is how this fits in with our strategy (past and future) for handling server-side, user-based experiments. did you look at any examples of how we've done these in the past, and is this approach similar to a particular past approach or possibly different due to the "split" nature of this feature?

> what if teachers need to go back to the old teacher dashboard?

You bring up a great point -- if I understand correctly, the current javascript-based experiment changes every link in both teacher dashboards to either point to pegasus or dashboard, is that correct? 

I think the idea here is that we would enable this for a small fraction of users (5%?) to limit the damage we do if there are any problems which we may have missed. if any of them are show-stoppers we can go back down to 0%, fix, go back up to 5%, and let it bake a while longer. so, the teachers wouldn't be able to get back to the old dashboard except by writing into zendesk to notify us of a blocking problem.

if we really want to give users a choice for a gradual transition, we could switch them all over to the new version, and put a "go back to old version" link somewhere. this might be a better user experience, but could require a bit more engineering investment.

**Comment by maddiedierker** [comment]
PTAL! i refactored the split test based on our conversation in stand-up this morning. now, the split test script will be run from both pegasus and dashboard teacher dashboards. i'm going to add a support link in the new teacher dashboard in a separate PR (per brad's suggestion)

**Review by davidsbailey — COMMENTED** [review]
Maddie and I talked about this approach at some length and it seems like it will work. Since I'm going to be out of office soon I want to share a few thoughts here:

### steps before launch
1. make sure we can see # of unique users on pegasus and dashboard versions of teacher dashboard via Google Analytics
2. manually verify that the kill switch of setting percent to zero works for each scenario

### limitations
* the split percentage, once set, cannot really be changed, other than to go down to 0%, or all the way up to 100%
*  individual users could see different teacher dashboards in different browsers/devices. this is probably ok, but means we shouldn't leave our users at a percentage besides 0% or 100% for very long.

### alternative

The best alternative is probably to create a  [UserBasedExperiment](https://github.com/code-dot-org/code-dot-org/blob/96ebca049e75921f8ad6a59a4965ca3a7d937a1e/dashboard/app/models/experiments/user_based_experiment.rb#L28) which sets an experiment for a % of users and can be changed at any time. Whether the user is in the experiment would be passed down to the client (via both dashboard and pegasus) and directly determine whether the user should be routed to the pegasus or dashboard version of teacher dashboard. There is a pegasus library for accessing these dashboard experiments. 

This approach would be simpler and wouldn't have the limitations listed above. But, to reiterate, if we think the approach currently in this PR will work too, then there may not be a compelling reason to do so.

### bottom line

I'm leaving this as "comments" rather than approval because I'd really like @islemaster and @madelynkasula  to agree on how to proceed, since I won't be around when this launches. Sorry to pass the buck. @islemaster , please review!

**Comment by maddiedierker** [comment]
thanks, @davidsbailey! re: this comment --

> the split percentage, once set, cannot really be changed, other than to go down to 0%, or all the way up to 100%

just to clarify, the split percentage is controlled by a DCDO flag (`teacher_dashboard_split_test`), so it can be set to any value 0-100 and it will affect the percentage of users opted-in to the experiment at any given time.

**Comment by davidsbailey** [comment]
That is true, but my understanding is that once a not zero percentage in set, users start getting bucketed into on or off, and that seeing gets persisted into Pegasus localstorage. So only those who did not visit at the old percentage will be bucketed at the new percentage. Is that correct?

**Comment by maddiedierker** [comment]
oh, yes, that's correct. does `UserBasedExperiment` work differently? my understanding was that once a user was bucketed into an experiment group, they stayed in that experiment group. so if the percentage associated with that experiment were to change, any users previously opted in/out of the experiment would still be in the same state, but new users would be opted in/out according to the new experiment percentage. 

i appreciate your patience on this! i just want to make sure i fully understand before i refactor or ship the change as-is

**Comment by davidsbailey** [comment]
The server side experiment would not be sticky. If you set it to 5%, then it is only sticky in the sense that each user with a `user_id % 100 < 5` will always be bucketed into the experiment. If you increase it to 10%, then those same users will still be in it, plus users with modulos 6...10.

**Comment by davidsbailey** [comment]
Sorry, fat finger.

Then if you lower it to zero, all users would be out of the experiment. 

To make the user experience reflect the server setting, we'd need to pipe the result of `UserBasedExperiment.enabled?` through to the client without caching it in local storage so that it is fresh every time the user loads the page.

**Comment by maddiedierker** [comment]
okay, i think i now fully understand the trade-offs/limitations between the two approaches. thank you for helping me understand! 

for the purposes of this experiment, it's okay if the experiment group is sticky because we want to push users to the new teacher dashboard (rather than actually A/B testing possible new functionality), and the split test percentage is more of a safe-guard against any bugs we missed.

**Comment by islemaster** [comment]
Not disagreeing with your diagram, but I did my own to make sure I understand all the flows through this system.

image

There was a surprise (to me) terminal node: `Old Y` is stable if the experiment is off.  This implies that even if we turn off the experiment, when we turn it back on it will come back on for all the users that were previously opted in.  This behavior seems okay to me, just wanted to make sure it's understood.

Are we going to actually associate any metrics with this, or are we satisfied that we'll hear about problems through zendesk or error reporting systems?

**Comment by maddiedierker** [comment]
@islemaster thank you for pointing that out! i knew that would be the behavior if we set the split test to 0 then turned it back on, but did a poor job of documenting it.

here are the metrics i planned on monitoring:
- unique users visiting old/new teacher dashboard (via Google Analytics)
- count of times the `teacher-dashboard-react` has been enabled (also Google Analytics). this metrics isn't very useful since it doesn't correspond to unique users, but is something i'd still like to keep an eye on.
- zendesk reports from users

let me know if you think there are other important metrics we should be monitoring!

**Review by islemaster — APPROVED** [review]
Scanning the conversation above it looks like you've considered everything that came to mind for me. This looks really good!  I appreciate the thought that's gone into how we might do this in the future too (even if it's not this appraoch).

**Comment by davidsbailey** [comment]
+1 for proceeding. Thanks all for taking time to think through the plan!
